---
title: "Tutorial for Stevens.scRNASeq"
author: "Chase Stevens"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial for Stevens.scRNASeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
  )
```

## Installation of package and dependencies

```{r inst, echo=T, results=T, message=F, warning=F,eval=FALSE}
# Install devtools package from CRAN
install.packages("devtools")

# Install package from Github
devtools::install_github(
  "cschasestevens/Stevens.scRNASeq", 
  ref = "master", 
  build_vignettes = T
  )

# Install presto package for faster performance of the Seurat FindAllMarkers() function
devtools::install_github("immunogenomics/presto")

# load package
library(Stevens.scRNASeq)

# Check to see if all required packages are installed; if not, install from CRAN and Bioconductor
sc.check.lib()

# Access package vignettes
??Stevens.scRNASeq
browseVignettes("Stevens.scRNASeq")

# Change working directory to analysis folder
## Please note, analysis folders should have the following subfolders prior to running functions: analysis/, ref/, data/, processed/
setwd("C:/Users/ncsteven/Documents/Analysis/tutorials.scRNASeq")
getwd()

## If running in WSL2, be sure to increase the memory limit by creating a .wslconfig file in the user directory (see online for specific details).
```


## Data processing

```{r data-proc, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

# Define dataset input parameters (replace with columns specific to study design)
list.params <- Create.proc.params(
  "data/",
  # Dataset-specific columns
  data.frame(
    # ID column name (splits by underscore, should be listed first in folder name as in 'D24S_LAE_KO_D28')
    "Code" = unlist(lapply(strsplit(basename(list.files("data/")),"_",fixed = T),"[",1)),
    # 1st metadata column (include in CellRanger folder name)
    "Knockout" = as.factor(ifelse(grepl("NG",basename(list.files("data/"))),"NG","KO")),
    )
  )

# Process single file
proc.data <- sc.process.file(
  # parameter list
  list.params,
  # sample ID
  1,
  # adj.rho proportion
  0.1,
  # minimum cells per feature
  5,
  # minimum features per cell
  200
  )

# Process batch
list.data <- sc.process.batch(
  # parameter list
  list.params,
  # adj.rho.proportion
  0.1,
  # min. cells/feature
  5,
  # min. feat/cell
  200,
  # Run in parallel?
  TRUE,
  # Proportion of cores to use (set to 1 if FALSE)
  0.5
  )

# Save data
saveRDS(
  list.data,
  "processed/test.data.rds"
  )

# Update parameter list
list.params <- dplyr::bind_rows(
  setNames(
    lapply(
      seq.int(
        1,
        length(list.data),
        1
        ),
      function(x)
        list.data[[x]][[2]]
      ),
    names(list.data)
    )
  )

# Plot adjusted rho for each sample
sc.plot.rho(list.params)

# Formatted integration list
list.integrate <- setNames(
  lapply(
    seq.int(
      1,
      length(list.data),
      1
      ),
    function(x)
      list.data[[x]][[1]]
      ),
  names(list.data)
  )

# Integrate data
d.integrated <- sc.integrate.data(
  list.integrate,
  TRUE,
  0.5
  )

# Integration QC
ggplot2::ggsave(
  "analysis/plot.integration.qc.png",
  sc.integration.qc(d.integrated,"seurat_clusters"),
  width = 24,
  height = 8,
  dpi = 700)


```

## PCA, UMAP, and Visualization of Gene Expression

```{r umap, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

# Load gene list
list.genes <- read.table(
  "ref/gene.list.txt",
  sep = "\t",
  header = T
  )
# Cell Type List (for Cell Type Prediction)
list.ct <- c(
  "Secretory",
  "SecretoryCiliated",
  "Ciliated",
  "Basal",
  "Suprabasal",
  "Cycling",
  "Ionocyte/NE"
  )
# Load integrated data
if(!exists("d.integrated")) {
  d.integrated <- readRDS("analysis/D28.integrated.analysis.annotated.rds")
  }
# Load reference data for annotating clusters
d.ref <- readRDS("ref/Ken_HBE10xLSAE5codes1mSAE_cls19noLQ15_seurat.rds")

# metadata columns for plot overlays
list.p.cols <- c(
  "Code","Airway",
  "Knockout"
  )

# Run PCA
d1 <- sc.pca(d.integrated)
d1.pca <- sc.pca.plot(
  d1,
  c("Code","Airway","Knockout","seurat_clusters")
  )
ggplot2::ggsave(
  "analysis/plot.pca.panel.png",
  d1.pca,
  width = 36,
  height = 12,
  dpi = 700)

# Determine optimal number of dimensions to use, then run UMAP
d1 <- Seurat::RunUMAP(
  object = d1,
  dims = 1:20,
  n.components = 3
  )
d1 <- SeuratObject::AddMetaData(
  d1,
  factor(
    d1@meta.data$seurat_clusters,
    levels = c(
      seq.int(
        0,
        length(
          unique(
            d1@meta.data$seurat_clusters
            )
          ) - 1,
          1
        )
      )
    ),
  col.name = "seurat.clusters"
  )
d1.umap1 <- sc.umap.panel(
  d1,
  c("seurat.clusters",list.p.cols)
  )
ggplot2::ggsave(
  "analysis/plot.umap.panel.png",
  d1.umap1,
  height = 16,
  width = 32,
  dpi = 400
  )


# Plot UMAPs of subsetted data
list.analysis.subs <- data.frame(
  "Comp.no" = c(1,2,3,4),
  "Airway" = c("SAE","SAE","LAE","LAE"),
  "Knockout" = c("Control","KO","Control","KO")
  )

list.analysis.subs.d <- setNames(
  lapply(
    list.analysis.subs[["Comp.no"]],
    function(x) {
      l.sub <- BiocGenerics::subset(
        d.integrated,
        subset = Airway == list.analysis.subs[x,"Airway"] &
          Knockout == list.analysis.subs[x,"Knockout"]
        )
      Seurat::DefaultAssay(l.sub) <- "RNA"
      return(l.sub)
    }
  ),
  c(
    paste(
      list.analysis.subs[["Airway"]],
      list.analysis.subs[["Knockout"]],
      sep = "."
    )
  )
)
# Create UMAPs of subsetted Seurat objects
list.analysis.subs.p <- ggpubr::ggarrange(
  plotlist = lapply(
    list.analysis.subs[["Comp.no"]],
    function(x) {

      d1.umap1 <- sc.umap.panel(
        list.analysis.subs.d[[x]],
        c("seurat.clusters",list.p.cols)
      )
      ggplot2::ggsave(
        paste(
          "analysis/plot.umap.panel.",
          list.analysis.subs[x,"Airway"],
          list.analysis.subs[x,"Knockout"],
          ".png",sep = ""),
        d1.umap1,
        height = 16,
        width = 32,
        dpi = 400
      )
    }
  ),
  nrow = 2,
  ncol = 2,
  common.legend = F)


# Create Gene expression UMAPs based on a specific gene list
sc.umap.panel.gene.list(
  # Gene list
  list.genes[[1]],
  # Seurat Object
  d.integrated,
  # Group variable
  "CellType",
  # Color scheme
  col.univ()[1:length(levels(d.integrated@meta.data[["CellType"]]))],
  # Color Names
  c(levels(d.integrated@meta.data[["CellType"]])),
  # legend x-position
  0.95,
  # legend y-position
  0.95,
  # Run in parallel? (set to FALSE if on Windows)
  TRUE,
  # Percentage of available cores to use
  0.5
  )

## Single plot
sc.umap.panel.gene(
  # Seurat Object
  d.integrated,
  # Group variable
  "CellType",
  # Gene name
  "CFTR",
  # Color scheme
  col.univ()[1:length(levels(d.integrated@meta.data[["CellType"]]))],
  # Color Names
  c(levels(d.integrated@meta.data[["CellType"]])),
  # legend x-position
  0.95,
  # legend y-position
  0.95
  )

```

## CellType Prediction

```{r cell-pred, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

# Automatic cell type prediction
gc(reset = T)
d.list <- sc.predict.clusters(
  # Input Seurat object
  d.integrated,
  # Reference Seurat object
  d.ref,
  # Future size of created objects (in MB)
  5000,
  # Cluster column name
  "seurat.clusters",
  # Vector of group metadata columns
  list.p.cols,
  # Run in parallel? (Set to FALSE if running on Windows)
  TRUE,
  # Proportion of available cores to use
  0.25
  )

saveRDS(d.list,"analysis/data.with.predictions.rds")

# Manually assign existing CellType and CellGroup columns
d.list.ct <- dplyr::select(
  dplyr::left_join(
    d.list$`Predicted Clusters`@meta.data["seurat.clusters"],
    data.frame(
      "seurat.clusters" = paste(sort(unique(d.list$`Predicted Clusters`@meta.data$seurat.clusters))),
      "CellType" = factor(
        c("0.Basal","1.Suprabasal","2.Secretory","3.SecretoryCiliated","4.Secretory",
          "5.SecretoryCiliated","6.Secretory","7.Ciliated","8.Ciliated","9.Ciliated",
          "10.Cycling","11.Secretory","12.SecretoryCiliated","13.Secretory","14.Cycling",
          "15.SecretoryCiliated","16.Secretory","17.SecretoryCiliated","18.Rare","19.Secretory",
          "20.Rare"),
        levels = c("0.Basal","1.Suprabasal","2.Secretory","3.SecretoryCiliated","4.Secretory",
                   "5.SecretoryCiliated","6.Secretory","7.Ciliated","8.Ciliated","9.Ciliated",
                   "10.Cycling","11.Secretory","12.SecretoryCiliated","13.Secretory","14.Cycling",
                   "15.SecretoryCiliated","16.Secretory","17.SecretoryCiliated","18.Rare","19.Secretory",
                   "20.Rare")),
      "CellGroup" = factor(
        c("Basal","Suprabasal","Secretory","SecretoryCiliated","Secretory",
          "SecretoryCiliated","Secretory","Ciliated","Ciliated","Ciliated",
          "Cycling","Secretory","SecretoryCiliated","Secretory","Cycling",
          "SecretoryCiliated","Secretory","SecretoryCiliated","Rare","Secretory",
          "Rare"),
        levels = c("Basal","Suprabasal","Cycling","Ciliated","SecretoryCiliated",
                   "Secretory","Rare"))
      ),
    by = "seurat.clusters"),
  .data[["CellType"]],.data[["CellGroup"]]
  )

d.list$`Predicted Clusters` <- SeuratObject::AddMetaData(
  d.list$`Predicted Clusters`,
  d.list.ct[["CellType"]],
  col.name = "CellType"
  )

d.list$`Predicted Clusters` <- SeuratObject::AddMetaData(
  d.list$`Predicted Clusters`,
  d.list.ct[["CellGroup"]],
  col.name = "CellGroup"
  )

## UMAP with final cluster assignments (Type and Group)
### Cell Type
ggplot2::ggsave(
  "analysis/plot.umap.CellType.png",
  sc.umap.standard(
    # Seurat object
    d.integrated,
    # metadata column
    "CellType"
    ),
  height = 8,
  width = 8,
  dpi = 700
  )
### Cell Group
ggplot2::ggsave(
  "analysis/plot.umap.CellGroup.png",
  sc.umap.standard(
    # Seurat object
    d.integrated,
    # metadata column
    "CellGroup"
    ),
  height = 8,
  width = 8,
  dpi = 700
  )

## Save analysis rds
saveRDS(d.list,"analysis/data.annotated.rds")


```


## Visualize Marker Genes

```{r mark-gene, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

## Create Heatmap
p.heatmap.top10 <- sc.top10.marker.heatmap(
  # Seurat object
  d.integrated,
  # Cluster column
  "CellType",
  # Heatmap width
  36,
  # Heatmap height
  12,
  # Column fontsize
  4,
  # Row fontsize
  8
  )

## Save
png(
  "analysis/plot.heatmap.top10.markers.png",
  width = 40,
  height = 14,
  units = "cm",
  res = 1000
  )
print(p.heatmap.top10)
dev.off()



```

## Analyze a Reclustered Subset of scRNA-Seq Data

```{r reclus, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

# Recluster based on two cell types
d.recluster <- sc.recluster.data(
  # Seurat object
  d.integrated,
  # Cell types for subsetting
  c("Secretory","SecretoryCiliated"),
  # gene list file (as vector)
  "ref/gene.list.txt",
  # metadata column list
  list.p.cols,
  # marker gene heatmap width
  36,
  # marker gene heatmap height
  12,
  # marker gene heatmap column fontsize
  6,
  # marker gene heatmap row fontsize
  8,
  # run in parallel? (Set to FALSE if on Windows)
  TRUE,
  # core percentage to use for plotting gene expression umaps
  0.5
  )


```

## Statistical Analysis

### Differential Gene Expression Analysis (DGEA)

```{r dgea, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

# Load gene list
list.genes <- read.table(
  "ref/gene.list.txt",
  sep = "\t",
  header = T
  )
# Load data
if(!exists("d.analysis")) {
  d.analysis <- readRDS("analysis/D28.integrated.analysis.annotated.rds")
  }
# metadata columns for plot overlays
list.p.cols <- c(
  "Code","Airway",
  "Knockout"
  )

# Subset and Run DGEA
d1 <- BiocGenerics::subset(
  d.analysis,
  subset = Airway == "SAE"
  )
remove(d.analysis)

dgea.output <- sc.DGEA(
  # Seurat object
  d1,
  # metadata column list
  c(list.p.cols,"CellType","nFeature_RNA"),
  # Cell type column name
  "CellType",
  # MAST comparison name
  "KnockoutKO",
  # MAST name (user-provided)
  "KO vs. NG ",
  # Formula
  as.formula(
    paste(
      "~","Knockout","+",
      "nFeature_RNA",
      sep = " "
      )
    ),
  # run in parallel? (Set to FALSE if on Windows)
  TRUE,
  # core percentage to use
  0.5
  )


```

### Visualize DGEA Results

```{r dgea-vis, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

if(!exists("dgea.output") & file.exists("analysis/object.dgea.result.rds")) {
  dgea.output <- readRDS("analysis/object.dgea.result.rds")
  }
# Top-10 DEG heatmap
p.heatmap <- sc.top10.deg.heatmap(
  # DGEA results list
  dgea.output,
  # Seurat object
  d1,
  # Cluster column name
  "CellType",
  # Heatmap width
  36,
  # Heatmap height
  12,
  # Heatmap column font size
  6,
  # Heatmap row font size
  8,
  # Comparison name
  "KO vs. NG"
  )

png(
  "analysis/plot.heatmap.top10.deg.KOvNG.png",
  width = 44,
  height = 14,
  units = "cm",
  res = 1600
  )
print(p.heatmap)
dev.off()


# DEG Dot Plot
p.dotplot <- sc.top10.deg.dotplot(
  # Use a custom gene list ('cstm.list') or dgea.results object ('deg.list')?
  "cstm.list",
  # DGEA results list or custom gene list
  list.genes,
  # Seurat object
  d1,
  # Pattern for matching to specific cell clusters (separate entries by |)
  "4.Secretory|13.Ion",
  # Clustering/CellType column name
  "CellType",
  # Group variable(s)
  c("Airway","Knockout")
  )

## Save
ggplot2::ggsave(
  "analysis/plot.dotplot.test.png",
  p.dotplot$Plot,
  width = 20,
  height = 8,
  dpi = 700
  )


# DEG Volcano Plot
## Single
p.vol <- sc.volcano(
  # DGEA results list or custom gene list
  dgea.output,
  # Cell type
  "2.Secretory",
  # Comparison name
  unique(dgea.output[[1]][["Comparison"]][[1]]),
  # P-value cutoff
  0.005,
  # Fold change cutoff
  0.25,
  # Fold change limit
  6
  )

ggplot2::ggsave(
  "analysis/plot.volplot.deg.KOvNG.2Sec.png",
  p.vol,
  width = 10,
  height = 8,
  dpi = 700
  )


```

## Gene Ontology/Custom Gene Set Enrichment Analysis

```{r go-enrich, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

# Run function (results are saved in 'analysis/' folder)
enr <- sc.enrichment(
  # Data frame containing DGEA results (created from sc.DGEA)
  l.deg = readRDS("analysis/object.dgea.result.rds")[[1]],
  # Enrichment type (Either "GO" or "cstm")
  en.type = "cstm",
  # If enrichment type is 'cstm', provide a custom gene set list (ignored when performing GO enrichment)
  cstm.list = setNames(read.table("ref/gene.list.lipids.txt",header = T,sep = "\t"),c("GENE","Set")),
  # run in parallel? (Set to FALSE if on Windows)
  parl = TRUE,
  # core percentage to use
  core.perc = 0.75
  )

# Custom gene set example
sc.plot.enrichment(
  enr[["Results"]][enr[["Results"]][["CellType"]] == "2.Secretory",],
  "cstm"
  )
# Gene ontology example
sc.plot.enrichment(
  enr[enr[["Description"]] == "GO_BP_4.Secretory",],
  "GO"
  )

```




