---
title: "tutorial_v2_0"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{tutorial_v2_0}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Installation of package and dependencies
```{r inst, echo=T, results=T, message=F, warning=F,eval=FALSE}
# Install devtools package from CRAN
install.packages("devtools")
# Install package from Github
devtools::install_github(
  "cschasestevens/Stevens.scRNASeq",
  ref = "master",
  build_vignettes = TRUE
)
# Install presto package for faster performance
# of the Seurat FindAllMarkers() function
devtools::install_github("immunogenomics/presto")
# load package
library(Stevens.scRNASeq)
# Check to see if all required packages are installed;
# if not, install from CRAN and Bioconductor
sc_check_lib()
# Access package vignettes
browseVignettes("Stevens.scRNASeq")
# Change working directory to analysis folder
getwd()
setwd("pathtofolder")
```


## Data processing
```{r data-proc, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# If processing scRNA-Seq only:
list_params <- create_proc_params(
  "data/",
  # Dataset-specific columns
  data.frame(
    # ID column name (splits by underscore, should be listed
    # first in folder name as in 'D24S_LAE_KO_D28')
    "Code" = unlist(
      lapply(
        strsplit(
          basename(list.files("data/")),
          "_",
          fixed = TRUE
        ),
        "[",
        1
      )
    ),
    # 1st metadata column (include in CellRanger folder name)
    "Knockout" = as.factor(
      ifelse(
        grepl(
          "NG",
          basename(list.files("data/"))
        ),
        "NG",
        "KO"
      )
    ),
  )
)

# If processing multiome data:
l_params <- sc_multiome_params(
  data.frame(
    # ID column name
    "Code" = basename(list.files("data/")),
    # 1st metadata column (include in CellRanger folder name)
    "Group" = as.factor(
      ifelse(
        grepl("group", basename(list.files("data/"))),
        "g1",
        "g2"
      )
    )
  ),
  "ref/gencode.v45.primary_assembly.annotation.gtf",
  "ref/GRCh38.primary_assembly.genome.fa"
)

# Process individual scRNA-Seq files
proc_data <- sc_process_file(
  # parameter list
  list_params,
  # sample ID
  1,
  # adj.rho proportion
  0.1,
  # minimum cells per feature
  5,
  # minimum features per cell
  200
)

# Process batch scRNA-Seq files
list_data <- sc_process_batch(
  # parameter list
  list_params,
  # adj.rho.proportion
  0.1,
  # min. cells/feature
  5,
  # min. feat/cell
  200,
  # Run in parallel?
  TRUE,
  # Proportion of cores to use (set to 1 if FALSE)
  0.5
)

# Process multiome files
list_data <- sc_process_batch_multiome(
  # parameter list
  l_params,
  # Project name
  "project1",
  # Minimum cells per feature
  5,
  # Minimum features per cell
  200,
  # ATAC counts upper limit
  100000,
  # RNA counts upper limit
  25000,
  # RNA counts lower limit
  1000,
  # Percentage mitochondrial reads upper limit
  20,
  # Nucleosome signal upper limit
  2.5,
  # TSS enrichment lower limit
  2,
  # Run in parallel?
  TRUE,
  # Core percentage
  0.1
)

# Save data
saveRDS(
  list_data,
  "processed/test.data.rds"
)

# Integrate data: either "scRNA-Seq" or "multiome"
d_integrated <- sc_integrate_data(
  l_so = list_data,
  l_par = l_params,
  proc_mode = "scRNA-Seq",
  parl = TRUE,
  core_perc = 0.5
)

# Integration QC
ggplot2::ggsave(
  "analysis/plot.integration.qc.png",
  sc_integration_qc(d_integrated, "seurat_clusters"),
  width = 24,
  height = 8,
  dpi = 700
)
```

## PCA, UMAP, and Visualization of Gene Expression
```{r umap, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Load gene list
list_genes <- read.table(
  "ref/gene.list.txt",
  sep = "\t",
  header = TRUE
)
# Cell Type List (for Cell Type Prediction)
list_ct <- c(
  "Secretory",
  "SecretoryCiliated",
  "Ciliated",
  "Basal",
  "Suprabasal",
  "Cycling",
  "Ionocyte/NE"
)
# Load integrated data
if(!exists("d_integrated")) { # nolint
  d_integrated <- readRDS("analysis/D28.integrated.analysis.annotated.rds")
}
# Load reference data for annotating clusters
d_ref <- readRDS("ref/Ken_HBE10xLSAE5codes1mSAE_cls19noLQ15_seurat.rds")

# metadata columns for plot overlays
list_p_cols <- c(
  "Code", "Airway", "Knockout"
)

# Run PCA
d1 <- sc_pca(d_integrated, "RNA")
d1_pca <- sc_pca_plot(
  d1,
  c(
    "Code", "Airway",
    "Knockout", "seurat_clusters"
  )
)
ggplot2::ggsave(
  "analysis/plot.pca.panel.png",
  d1.pca,
  width = 36,
  height = 12,
  dpi = 700
)

# Determine optimal number of dimensions to use, then run UMAP
d1 <- Seurat::RunUMAP(
  object = d1,
  dims = 1:20,
  n.components = 3
)
d1 <- SeuratObject::AddMetaData(
  d1,
  factor(
    d1@meta.data$seurat_clusters,
    levels = c(
      seq.int(
        0,
        length(
          unique(
            d1@meta.data$seurat_clusters
          )
        ) - 1,
        1
      )
    )
  ),
  col.name = "seurat.clusters"
)
d1_umap1 <- sc_umap_panel(
  # Seurat Object
  d1,
  # Metadata columns to overlay
  c("col1", "col2", "col3"),
  # Slot to use for dimension reduction
  "wnn.umap"
)
ggplot2::ggsave(
  "analysis/plot.umap.panel.png",
  d1_umap1,
  height = 16,
  width = 32,
  dpi = 400
)

# Create Gene expression UMAPs based on a specific gene list
sc_umap_panel_gene_list(
  # Gene list
  list_genes[[1]],
  # Seurat Object
  d1,
  # Group variable
  "CellType",
  # Color scheme
  col_univ()[1:length(levels(d1@meta.data[["CellType"]]))], # nolint
  # Color Names
  c(levels(d1@meta.data[["CellType"]])),
  # legend x-position
  0.95,
  # legend y-position
  0.95,
  # Run in parallel? (set to FALSE if on Windows)
  TRUE,
  # Percentage of available cores to use
  0.5
)

## Single plot
sc_umap_panel_gene(
  # Seurat Object
  d1,
  # Group variable
  "CellType",
  # Gene name
  "CFTR",
  # Color scheme
  col_univ()[1:length(levels(d1@meta.data[["CellType"]]))], # nolint
  # Color Names
  c(levels(d1@meta.data[["CellType"]])),
  # legend x-position
  0.95,
  # legend y-position
  0.95
)
```

## CellType Prediction
```{r cell-pred, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Seurat cell type prediction
gc(reset = TRUE)
d_list <- sc_predict_clusters(
  # Input Seurat object
  d1,
  # Reference Seurat object
  d_ref,
  # Future size of created objects (in MB)
  5000,
  # Cluster column name
  "seurat.clusters",
  # Vector of group metadata columns
  list_p_cols,
  # Run in parallel? (Set to FALSE if running on Windows)
  TRUE,
  # Proportion of available cores to use
  0.25
)
# Azimuth cell type prediction
library(Azimuth)
options(future.globals.maxSize = 10000 * 1024^2)
d_list <- Azimuth::RunAzimuth(
  d_int,
  reference = "lungref"
)
### Save
write.table(
  d_list@meta.data,
  "analysis/table.predicted.cell.types.txt",
  row.names = FALSE,
  col.names = TRUE,
  sep = "\t"
)
saveRDS(d_list, "analysis/data.predictedtype.rds")

# Manually assign existing CellType and CellGroup columns
# using marker genes and prediction scores
h_out <- sc_top10_marker_heatmap(
  # Seurat object
  d_pred,
  # Assay to use
  "RNA",
  # Cluster column
  "seurat_clusters",
  # Heatmap width
  36,
  # Heatmap height
  12,
  # Column fontsize
  4,
  # Row fontsize
  8
)
png(
  "analysis/plot.heatmap.top10.markers.png",
  width = 42,
  height = 18,
  units = "cm",
  res = 1000
)
print(h_out)
dev.off()

## seurat_cluster UMAP
ggplot2::ggsave(
  "analysis/plot.umap.seuratclusters.png",
  sc_umap_standard(
    # Seurat object
    d_list,
    # metadata column
    "seurat_clusters",
    # reduction to plot
    "wnn.umap"
  ),
  height = 8,
  width = 8,
  dpi = 700
)

## Save analysis rds
saveRDS(d_list, "analysis/data.annotated.rds")
```

## Analyze a Reclustered Subset of scRNA-Seq Data

```{r reclus, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Recluster based on two cell types
d_recluster <- sc_recluster_data(
  # Seurat object
  d_list,
  # Cell types for subsetting
  c("Secretory", "SecretoryCiliated"),
  # gene list file (as vector)
  "ref/gene.list.txt",
  # metadata column list
  list_p_cols,
  # marker gene heatmap width
  36,
  # marker gene heatmap height
  12,
  # marker gene heatmap column fontsize
  6,
  # marker gene heatmap row fontsize
  8,
  # run in parallel? (Set to FALSE if on Windows)
  TRUE,
  # core percentage to use for plotting gene expression umaps
  0.5
)
```

## Statistical Analysis
### Differential Gene Expression/Accessibility Analysis
```{r dgea, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# DGEA
dgea_output <- sc_diff(
  # Seurat object
  d,
  # Assay
  "RNA",
  # Slot
  "data",
  # metadata column list
  c("Code", "Airway", "CellType", "nFeature_RNA"),
  # Cell type column name
  "CellType",
  # MAST comparison name
  "AirwaySAE",
  # MAST name (user-provided)
  "SAE vs. LAE",
  # Formula
  as.formula(
    paste(
      "~", "Airway", "+",
      "nFeature_RNA",
      sep = " "
    )
  ),
  # run in parallel? (Set to FALSE if on Windows)
  TRUE,
  # core percentage to use
  0.5,
)

# Differential accessibility for annotated transcription factors
# (Adapted from Seurat)
## load data
d <- readRDS("analysis/data.annotated.rds")[[1]]

ref_gene1 <- rtracklayer::import(
  "ref/gencode.v45.primary_assembly.annotation.gtf"
)

## Load libraries
library(BSgenome.Hsapiens.UCSC.hg38)
library(chromVAR)
library(JASPAR2020)
library(TFBSTools)

## Add motifs
d_chrmasy <- sc_atac_motifs(
  d,
  ref_gene1
)

#### Differential Activity Analysis ####
## Add chromassay to Seurat
d[["TF"]] <- d_chrmasy
Seurat::DefaultAssay(d) <- "TF"
Seurat::Idents(d) <- "CellType"
diff.output.tf <- Seurat::FindAllMarkers(
  d,
  min.pct = 0.05,
  verbose = TRUE
)
d <- Signac::RegionStats(
  d,
  genome = BSgenome.Hsapiens.UCSC.hg38
)

## Motif enrichment
tf_enrich <- Signac::FindMotifs(
  object = d,
  features = rownames(d),
  assay = "TF"
)

write.table(
  tf.enrich,
  file = "analysis/table.enrich.tf.txt",
  col.names = TRUE,
  row.names = FALSE,
  sep = "\t"
)

## Motif activities
d <- Signac::RunChromVAR(
  object = d,
  genome = BSgenome.Hsapiens.UCSC.hg38
)
Seurat::DefaultAssay(d) <- "chromvar"
Seurat::Idents(d) <- "CellType"

diff_output_activity <- Seurat::FindAllMarkers(
  d,
  min.pct = 0.05,
  verbose = TRUE,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

diff_output_activity[["gene"]] <- unlist(
  lapply(
    row.names(diff_output_activity),
    function(x) {
      name(TFBSTools::getMatrixByID(JASPAR2020, ID = x))
    }
  )
)
diff_output_activity[["ID"]] <- rownames(diff_output_activity)

write.table(
  diff_output_activity,
  file = "analysis/table.diff.tf.celltype.txt",
  sep = "\t",
  col.names = TRUE,
  row.names = FALSE
)
saveRDS(d, "analysis/data.annotated.withTFs.rds")

```

### Visualize DGEA/DA Results
```{r dgea-vis, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}
# Differential Accessibility
tf_heatmap <- sc_top_motif_heatmap(
  # Seurat object
  so = d,
  # Differential activity results
  mot_m = diff_output_activity,
  # Filter based on CellType?
  filt_ct = TRUE,
  # Cell type name
  ct_name = "ct1",
  # Clustering column
  cl_var = "CellType",
  # Number of motifs to use
  top_n = 100,
  # Heatmap width
  h_w = 36,
  # Heatmap height
  h_h = 12,
  # Column font size
  fs_c = 4,
  # Row font size
  fs_r = 8
)

png(
  "analysis/plot.heatmap.motif.Top50.ct1.png",
  width = 40,
  height = 14,
  units = "cm",
  res = 1000
)
print(tf_heatmap)
dev.off()

p_vol <- sc_volcano(
  l_deg = diff_output_activity,
  comp_name = "grp2 vs. grp1 TF Motifs",
  gene_name = "near.gene",
  diff_col = "avg_log2FC",
  p_col = "p_val_adj",
  p_cut = 5e-3,
  f_cut = 0.35,
  f_lim = 4,
  y_limit = 300,
  x_title = paste(
    "Higher Activity in group1",
    "<----->",
    "Higher Activity in group2",
    sep = " "
  )
)

ggplot2::ggsave(
  paste(
    "analysis/",
    "plot.vol.marker.motif",
    ".png",
    sep = ""
  ),
  p_vol,
  height = 8,
  width = 8,
  dpi = 700
)

# DGEA
p_heatmap <- sc_top10_de_da_heatmap(
  # DGEA results list
  dgea_output,
  # Seurat object
  d,
  # Assay
  "RNA",
  # Cluster column name
  "CellType",
  # Heatmap width
  36,
  # Heatmap height
  12,
  # Heatmap column font size
  6,
  # Heatmap row font size
  8,
  # Comparison name
  "grp2 vs. grp1"
)

png(
  "analysis/plot.heatmap.top10.deg.png",
  width = 44,
  height = 14,
  units = "cm",
  res = 1600
)
print(p_heatmap)
dev.off()

# DEG Dot Plot
p_dotplot <- sc_top10_deg_dotplot(
  # Type "deg.list" or "cstm.list" to toggle between inputs
  "deg.list",
  # Name of a custom gene list or dgea.results object
  dgea_output,
  # Seurat object
  d_annotated,
  # Unique character strings corresponding to cell types
  "3.Se|6.Se",
  # Name of clustering variable
  "CellType",
  # Vector of up to 2 variables for stratifying clustering variables
  c("Knockout", "Airway")
)

## Save
ggplot2::ggsave(
  "analysis/plot.dotplot.png",
  p_dotplot$Plot,
  width = 20,
  height = 8,
  dpi = 700
)

# DEG Volcano Plot
p_vol <- sc_volcano(
  # DGEA results list or custom gene list
  dgea_output,
  # Cell type
  "ct1",
  # Comparison name
  unique(diff.output[[1]][["Comparison"]][[1]]),
  # P-value cutoff
  0.005,
  # Fold change cutoff
  0.25,
  # Fold change limit
  6,
  # y-axis limit
  50
)

ggplot2::ggsave(
  "analysis/plot.volplot.deg.ct1.png",
  p_vol,
  width = 10,
  height = 8,
  dpi = 700
)
```

## Gene Ontology/Custom Gene Set Enrichment Analysis **(Development)**
```{r go-enrich, echo=T, fig.align= "center", out.width="90%", results=T, message=F, warning=F,eval=FALSE}

# Run function (results are saved in 'analysis/' folder)
enr <- sc_enrichment(
  # Data frame containing DGEA results (created from sc.DGEA)
  l_deg = readRDS("analysis/object.dgea.result.rds")[[1]],
  # Enrichment type (Either "GO" or "cstm")
  en_type = "cstm",
  # If enrichment type is 'cstm', provide a custom gene
  # set list (ignored when performing GO enrichment)
  cstm_list = setNames(
    read.table(
      "ref/gene.list.lipids.txt",
      header = TRUE,
      sep = "\t"
    ),
    c("GENE", "Set")
  ),
  # run in parallel? (Set to FALSE if on Windows)
  parl = TRUE,
  # core percentage to use
  core_perc = 0.75
)

# Custom gene set example
sc_plot_enrichment(
  enr[["Results"]][enr[["Results"]][["CellType"]] == "ct1", ],
  "cstm"
)
# Gene ontology example
sc_plot_enrichment(
  enr[enr[["Description"]] == "GO_BP_4.Secretory", ],
  "GO"
)
```
